nixops/util.py

```diff
@@ -143,30 +146,37 @@ def __init__(self, *args: ImmutableValidatedObject, **kwargs):
                 continue
             anno.update(x.__annotations__)
 
-        def _transform_value(key: Any, value: Any) -> Any:
-            ann = anno.get(key)
-
+        def _transform_value(value: Any, target_type: Optional[Type]) -> Any:
             # Untyped, pass through
-            if not ann:
+            if not target_type:
                 return value
 
-            if inspect.isclass(ann) and issubclass(ann, ImmutableValidatedObject):
-                value = ann(**value)
-
-            # Support Sequence[ImmutableValidatedObject]
-            if isinstance(value, tuple) and not isinstance(ann, str):
-                new_value = []
-                for v in value:
-                    for subann in ann.__args__:
-                        if inspect.isclass(subann) and issubclass(
-                            subann, ImmutableValidatedObject
-                        ):
-                            new_value.append(subann(**v))
-                        else:
-                            new_value.append(v)
-                value = tuple(new_value)
-
-            typeguard.check_type(key, value, ann)
+            # Support ImmutableValidatedObject
+            if (
+                isinstance(value, Mapping)
+                and inspect.isclass(target_type)
+                and issubclass(target_type, ImmutableValidatedObject)
+            ):
+                value = target_type(**value)
+
+            type_origin = typing.get_origin(target_type)  # type: ignore[attr-defined]
+            type_args = tuple(set(typing.get_args(target_type)) - {type(None)})  # type: ignore[attr-defined]
+            if (
+                type_origin is not None
+                and len(type_args) == 1
+                and inspect.isclass(type_args[0])
+                and issubclass(type_args[0], ImmutableValidatedObject)
+            ):
+                # Support Sequence[ImmutableValidatedObject]
+                if isinstance(value, Sequence) and issubclass(tuple, type_origin):
+                    value = tuple(_transform_value(v, type_args[0]) for v in value)
+
+                # Support Optional[ImmutableValidatedObject]
+                if type_origin is Union:
+                    if value is not None:
+                        value = _transform_value(value, type_args[0])
```

Comment:
I discovered that the logic for this is slightly off, so hold off reviewing for now please.